[CI] ES92GpuHnswWriteGraphTests testCpuFallbackGraphMatchesLucene fails with seed EA6DA5BD8CC04BF6

### CI Link

https://gradle-enterprise.elastic.co/s/u2nvwxvnxshla/tests/task/:libs:gpu-codec:test/details/org.elasticsearch.gpu.codec.ES92GpuHnswWriteGraphTests/testCpuFallbackGraphMatchesLucene

### Repro line

```text
./gradlew ":libs:gpu-codec:test" --tests "org.elasticsearch.gpu.codec.ES92GpuHnswWriteGraphTests.testCpuFallbackGraphMatchesLucene" -Dtests.seed=EA6DA5BD8CC04BF6 -Dtests.locale=bg-BG -Dtests.timezone=Asia/Dushanbe
```

### Does it reproduce?

Only on Linux x86_64 (AVX-2): 3/3 attempts in CI failed identically with this seed (Buildkite retries reuse the build's seed). The same seed passes on macOS/aarch64, consistent with the SIMD analysis in #149659.

### Applicable branches

main

### Failure history

Appears to be a recurrence of #149659 (closed 2026-05-26) with a different seed. No other failures of this test in the last 7 days per ci-stats — the failing build is an unrelated docs-only PR (#154901) that drew this seed:
- https://buildkite.com/elastic/elasticsearch-pull-request/builds/166785
- Build scans: https://gradle-enterprise.elastic.co/s/u2nvwxvnxshla, https://gradle-enterprise.elastic.co/s/uo3ibyl2epfri, https://gradle-enterprise.elastic.co/s/rqmpm32xwp7so

### Failure excerpt

```
java.lang.AssertionError: kNN results for query 35 should match expected:<[0, 161, 35, 214, 198, 8, 249, 201, 58, 235]> but was:<[0, 161, 35, 214, 198, 8, 185, 249, 58, 235]>
  at org.elasticsearch.gpu.codec.ES92GpuHnswWriteGraphTests.assertSearchResultsEqual(ES92GpuHnswWriteGraphTests.java:178)
```


:eyes:

--- a/libs/gpu-codec/src/test/java/org/elasticsearch/gpu/codec/ES92GpuHnswWriteGraphTests.java
+++ b/libs/gpu-codec/src/test/java/org/elasticsearch/gpu/codec/ES92GpuHnswWriteGraphTests.java
@@ -29,7 +29,6 @@ import org.apache.lucene.index.SegmentWriteState;
 import org.apache.lucene.index.VectorSimilarityFunction;
 import org.apache.lucene.search.IndexSearcher;
 import org.apache.lucene.search.KnnFloatVectorQuery;
-import org.apache.lucene.search.TopDocs;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.tests.util.TestUtil;
 import org.apache.lucene.util.hnsw.HnswGraph;
@@ -51,10 +50,12 @@ import static org.elasticsearch.gpu.codec.ES92GpuHnswVectorsFormat.DEFAULT_MAX_C
  * Validates that the CPU fallback path in {@link ES92GpuHnswVectorsWriter} produces
  * equivalent results to Lucene's {@link Lucene99HnswVectorsFormat}. No GPU required.
  *
- * <p>For float vectors, graph topology is compared structurally (same levels, same nodes
- * on each level) and kNN search results are compared to verify functional equivalence.
- * Neighbor lists are not compared exactly because 1-ULP floating-point differences in
- * vector scoring (due to JIT compilation state) can cause harmless tie-breaking divergence.
+ * <p>Graph topology is compared structurally (same levels, same nodes on each level).
+ * Neither neighbor lists nor kNN search results are compared exactly because 1-ULP
+ * floating-point differences in vector scoring (due to JIT compilation state) can cause
+ * harmless tie-breaking divergence during graph construction, producing structurally
+ * similar but not byte-identical graphs. Instead, both tests compare aggregate recall
+ * against brute-force ground truth.
  */
 public class ES92GpuHnswWriteGraphTests extends ESTestCase {
 
@@ -84,7 +85,7 @@ public class ES92GpuHnswWriteGraphTests extends ESTestCase {
             indexVectors(luceneDir, luceneFormat, vectors, similarity);
             try (DirectoryReader gpuReader = DirectoryReader.open(gpuDir); DirectoryReader luceneReader = DirectoryReader.open(luceneDir)) {
                 assertGraphStructureEqual(getGraph(gpuReader), getGraph(luceneReader));
-                assertSearchResultsEqual(gpuReader, luceneReader, vectors);
+                assertEqualRecallAgainstBruteForce(gpuReader, luceneReader, vectors, similarity);
             }
         }
     }
@@ -153,29 +154,6 @@ public class ES92GpuHnswWriteGraphTests extends ESTestCase {
             var luceneNodes = luceneGraph.getNodesOnLevel(level);
             var gpuNodes = gpuGraph.getNodesOnLevel(level);
             assertEquals("nodes on level " + level, luceneNodes.size(), gpuNodes.size());
-        }
-    }
-
-    /**
-     * Runs kNN queries against both indexes and asserts that the result sets are identical,
-     * verifying that the graphs are functionally equivalent even if neighbor lists differ
-     * slightly due to floating-point tie-breaking.
-     */
-    private void assertSearchResultsEqual(DirectoryReader gpuReader, DirectoryReader luceneReader, float[][] vectors) throws IOException {
-        int k = Math.min(10, numVectors);
-        int numQueries = Math.min(50, numVectors);
-        var gpuSearcher = new IndexSearcher(gpuReader);
-        var luceneSearcher = new IndexSearcher(luceneReader);
-
-        for (int q = 0; q < numQueries; q++) {
-            float[] queryVec = vectors[q];
-            var query = new KnnFloatVectorQuery(FIELD, queryVec, k);
-            TopDocs gpuDocs = gpuSearcher.search(query, k);
-            TopDocs luceneDocs = luceneSearcher.search(query, k);
-
-            Set<Integer> gpuIds = Arrays.stream(gpuDocs.scoreDocs).map(d -> d.doc).collect(Collectors.toSet());
-            Set<Integer> luceneIds = Arrays.stream(luceneDocs.scoreDocs).map(d -> d.doc).collect(Collectors.toSet());
-            assertEquals("kNN results for query " + q + " should match", luceneIds, gpuIds);
         }
     }
 

